### ISSUE TITLE
IP Access Control should retain data of IP Account owner

### ISSUE THREAD

**Issue by jdubpark** [storyprotocol/protocol-core-v1#59]
## Issue

The current IP Access Control mapping's key is
```solidity
address ipAccount => address signer => address to => bytes4 func
```

This presents an attack case where:
1. Alice owns IP 1, which contains tokens
2. Alice permissions her malicious "signer" to call all functions of TokenWithdrawalModule on behalf of IP 1
3. Alice sells IP 1 to Bob (selling the tokens in it together)
4. Right after Bob receives IP 1, Alice withdraws all tokens in IP 1 using the malicious signer

Hence, there should be an option to invalidate permissions on IP NFT transfers automatically.

## Simple Solution
As a naive solution, we can maintain another field in the mapping, as such
```solidity
address ipOwner => address ipAccount => address signer => address to => bytes4 func
```

Then, in `_encodePermission()`, the current owner of IP NFT would be used to encode the permission mapping.

In this case, step (3) would fail as the encoded permission now uses Bob's address (owner) in the encoding, which returns an unset permission.

## Advanced Solution

The solution above means transfers to unique owners invalidate existing permissions. This might not be ideal for some users who want to keep the old permissions set by previous owners, so that they don't have to set permissions again.

In other words, we want to facilitate the reuse of permissions between unique owner addresses. This will require a few more mappings and worse gas performance. 

Here's a rough sketch (might or might not work neatly):

1. Replace `address ipOwner` with `uint16 index`.
When checking permission, `index` would increment if the last checked owner is different from the current owner. Thus, after Alice transfers IP 1 to Bob, this `index` will increase by 1 (when any party attempts to interact with the protocol).

2. IP owner can choose to use any `index`.
If Bob wants to use all permissions of Alice, then he can choose to use Alice's `index`.

**Comment by kingster-will** [comment]
This feature has already been addressed since the Beta.

**Comment by jdubpark** [comment]
Can you elaborate on where? Please include relevant code, thanks

**Comment by kingster-will** [comment]
<img width="957" alt="Screenshot 2024-04-12 at 17 36 19" src="https://github.com/storyprotocol/protocol-core-v1/assets/83567446/6cae28ca-136d-4b95-a04b-42c5e017fd46">

The owner is encoded in the permission.

**Comment by jdubpark** [comment]
Thanks, so I missed the first param in encoding that provides the "simple solution."

What about
> The solution above means transfers to unique owners invalidate existing permissions. This might not be ideal for some users who want to keep the old permissions set by previous owners, so that they don't have to set permissions again.

**Comment by kingster-will** [comment]
I understand. This could be a use case focused on 'ease of use,' although it may not be a core functionality. Therefore, it might be best to consider utilizing application-level or peripheral contracts to support this scenario if necessary.

**Comment by jdubpark** [comment]
I don't think we should delegate core ACL to applications, but rather support in the core protocol. I agree it's more ease of use but might catch some users off-guard (thinking that old access remains the same so they don't need to configure).
